ports/stm32: Add LEGO Hub No. 7.

dlech opened the pull request:
This adds support for the LEGO Hub No. 7, aka LEGO Technic Small hub, aka LEGO SPIKE Essential hub.  This board is largely similar to Hub No. 6:

- Same MCU (STM32F413 - different packaging with fewer pins).
- Same Bluetooth chip (TI CC2564).
- Same IMU chip.
- Similar external flash chip - 4MiB instead of 32MiB.
- 2 I/O ports instead of 6.
- No display - only status and battery LEDs.
- Different LED driver chip.
- Only 1 button which is also the power button.
- No speaker.

dlech committed:
ports/stm32: Add LEGO Hub No. 7.

This adds support for the LEGO Hub No. 7, aka LEGO Technic Small hub,
aka LEGO SPIKE Essential hub.  This board is largely similar to Hub No.
6:

- Same MCU (STM32F413 - different packaging with fewer pins).
- Same Bluetooth chip (TI CC2564).
- Same IMU chip.
- Similar external flash chip - 4MiB instead of 32MiB.
- 2 I/O ports instead of 6.
- No display - only status and battery LEDs.
- Different LED driver chip.
- Only 1 button which is also the power button.
- No speaker.

Signed-off-by: David Lechner <david@pybricks.com>

dlech committed:
ports/stm32/LEGO_HUB_NO6/cc2564: Make timer configurable.

This adds configurable macros to define the timer and channel used to
provide the Bluetooth 32768 MHz clock. This will allow code to be
shared with LEGO_HUB_NO7.

Signed-off-by: David Lechner <david@pybricks.com>

dlech commented:
I have tested flashing MicroPython using DFU running from mboot and the bluetooth example from readme so this is at least mostly working.

I haven't been able to get the file system to mount yet though, so maybe I missed some difference on the external flash?

dlech commented:
I also didn't have any good ideas yet on how to have a bootloader UI with only one button that is also the power button. Maybe we just skip the UI and only go into bootloader mode with `machine.bootloader()`?

laurensvalk commented on ports/stm32/boards/LEGO_HUB_NO7/mpconfigboard.h:
```diff
@@ -0,0 +1,156 @@
+#define MICROPY_HW_SPIFLASH_BLOCKMAP(bl)         ((bl) + MICROPY_HW_SPIFLASH_OFFSET_BYTES / FLASH_BLOCK_SIZE)
+#define MICROPY_HW_SPIFLASH_BLOCKMAP_EXT(bl)     ((bl) + MICROPY_HW_SPIFLASH_OFFSET_BYTES / MP_SPIFLASH_ERASE_BLOCK_SIZE)
+#define MICROPY_HW_SPIFLASH_ENABLE_CACHE         (1)
+#define MICROPY_HW_SPIFLASH_SIZE_BITS            (32 * 1024 * 1024 - MICROPY_HW_SPIFLASH_OFFSET_BYTES * 8)
```
32 => 4?

laurensvalk commented:
> I haven't been able to get the file system to mount yet though, so maybe I missed some difference on the external flash?

It starts in 3-byte address mode.

dpgeorge commented:
> It starts in 3-byte address mode

Actually, it looks like the SPI flash doesn't support the 4-byte address mode commands (because it's small).  It doesn't matter what mode it starts in (3 or 4 byte) if `MICROPY_HW_SPI_ADDR_IS_32BIT` is forced to 0 or 1.

dlech commented:
> I also didn't have any good ideas yet on how to have a bootloader UI

We did some brainstorming and came up with a solution. Since the button is not usable for such purposes, we can use the presence/absence of the battery as the input to activate the mboot UI. The exact procedure is detailed in the README.

dlech commented on ports/stm32/boards/LEGO_HUB_NO7/README.md:
```diff
@@ -0,0 +1,135 @@
+different colors. Replace the battery (the button will not work without the
+battery present). Press the button to activate the desired boot mode:
+
+- Status light is red - run application (normal boot).
+- Status light is green - run application in factory file system mode.
+- Status light is blue - run application in safe mode.
+- Status light is white - start DFU on the USB port.
```
Are there standard colors for the bootloader modes? I found something that says white is DFU mode on one of the Pyboards, but that was all.

dpgeorge replied:
Yes, white is used on PYBD.

dpgeorge commented:
There are quite a few files here that are copied across from LEGO_HUB_NO6.  It would be good to eliminate duplication.

Eg:
- `bluetooth_init_cc2564C_1.5.c`, it could just `#include "../LEGO_HUB_NO6/bluetooth_init_cc2564C_1.5.c"`
- `cc2564.c` could be made the same for both boards by adding config values for the timer number and timer channel number
- `bdev.c` is the same and could also just use `#include`
- `spiflash.py` is the same and the `manifest.py` could just refer to the one copy
- `appupdate.py` is almost the same, it could be made to work on both by using the value of `sys.implementation._machine`, to detect which board it's on

dlech commented:
I have made the requested changes and rebased. 

I have to admit, though, that the runtime check in the .py files feels a bit fragile, e.g. what if someone decides to tweak the board name?

dlech committed:
ports/stm32/boards/LEGO_HUB_NO6: Fix typo in README.

This fixes a typo in the build directory path.

Signed-off-by: David Lechner <david@pybricks.com>

dpgeorge commented:
> I have to admit, though, that the runtime check in the .py files feels a bit fragile, e.g. what if someone decides to tweak the board name?

Some ideas for this:
1. add an alias like `FLASH_SCK` for `PB13`/`PB10` to `pins.csv`
2. add a method to `spiflash.py` to detect flash size (query the id register of the SPI flash)
3. use `pyb.Flash(start=0)` object to detect the flash size

dlech committed:
ports/stm32/boards/LEGO_HUB_NO6: Use named pins.

This changes all uses of pins to use the alias names of the pins. This
makes the code easier to understand and will also allow sharing more
code with LEGO_HUB_NO7.

Signed-off-by: David Lechner <david@pybricks.com>

dlech committed:
ports/stm32/LEGO_HUB_NO6/appupdate: Get filesystem size at runtime.

This changes appupdate.py to get the filesystem size at runtime. This
will allow the code to be shared with LEGO_HUB_NO7 which has a similar
flash chip with a different size.

Signed-off-by: David Lechner <david@pybricks.com>

dlech committed:
ports/stm32/LEGO_HUB_NO6/spiflash: Pick command type at runtime.

This changes spiflash.py to read the flash chip ID at runtime to select
the read/write/erase commands. This will allow the code to be shared
with LEGO_HUB_NO7 which doesn't use the 32-bit commands.

Also remove an unused constant while we are touching this.

Signed-off-by: David Lechner <david@pybricks.com>

dlech commented:
I implemented the suggested ideas (thanks!) and also dropped the duplicated linker files.

I've run through a bluetooth check and app update on both hubs to make sure everything is working.

dpgeorge commented on ports/stm32/boards/LEGO_HUB_NO6/spiflash.py:
```diff
@@ -4,18 +4,34 @@
+        self._WRITE = _CMD_WRITE_32 if _32_bit else _CMD_WRITE
+        self._ERASE = _CMD_SEC_ERASE_32 if _32_bit else _CMD_SEC_ERASE
+
+    def _get_id(self):
```
With this implemented, I think you can extract the SPI flash size from it.  That could be used instead of `pyb.Flash(start=0)`.  (This is only an observation, no need to change anything, unless you want to.)

dpgeorge commented on ports/stm32/boards/LEGO_HUB_NO7/README.md:
```diff
@@ -0,0 +1,135 @@
+Before installing MicroPython it is advised to backup the original LEGO firmware that
+the Hub comes installed with.  To do this, enter the built-in bootloader by holding
+down the Bluetooth button for 5 seconds while powering up the Hub via USB (you may
+need to take out the battery and disconnect USB to power off the Hub first).  Then
```
I think these instructions to enter the built-in bootloader are different on HUB_NO7.  There is no Bluetooth button.

dpgeorge commented:
Thanks for eliminating the duplication, it's looking a lot better.  Thanks also for testing the app update.  I tested mboot and BLE and they work.

dlech commented on ports/stm32/boards/LEGO_HUB_NO7/README.md:
```diff
@@ -0,0 +1,135 @@
+    $ make BOARD=LEGO_HUB_NO7
+
+Now enter the built-in bootloader by holding down the Bluetooth button for 5
+seconds while powering up the Hub via USB (you may need to take out the battery
```
fixed

dpgeorge commented:
Thanks for updating.  Merged in 6152bbe3dd82fb7dc1b063e24c19a42af03bf321 through 10f85fee183404a3cba2795e6cd22c12be742fb3
